### DISCUSSION TITLE
Shouldn't method addresses be implicitly convertible to void* and thus allowing direct casts to function pointers with mismatched signatures?

### DISCUSSION THREAD

**Discussion by Sergio0694** [dotnet/csharplang#5720] [Language Ideas] [answered]
The way the casting rules for function pointers work today make it extremely clunky and verbose to get a function pointer to a method and then casting it to one with a mismatched signature, forcing developers to either write two casts (super verbose) or split the declaration out into multiple temporary locals to avoid having very hard to read statements. Consider this example:

```csharp
unsafe
{
    // Ok, but looks ugly
    var f1 = (delegate*<uint, void>)(delegate*<int, void>)&Foo<int>;
    
    // Build error (why?)
    var f2 = (delegate*<uint, void>)&Foo<int>;
}

static void Foo<T>(T x)
{
}
```

The question is: in cases where `&Method` is already definitely resolved to be a specific method overload without ambiguities, shouldn't the language treat the returned expression to be of the right function pointer type directly, and given that is a subset of `void*`, then allow direct casts from it to any other pointer type, including to function pointers with mismatched signatures?

This would greatly simplify code doing things like this, and make it less error prone as well.

cc. @tannergooding @333fred

**Comment by CyrusNajmabadi** [comment]
That seems like it would be very error prone. It's much rather it forces you to be explicit that you are converting in this fashion.

**Reply by tannergooding** [reply]
I think the confusing thing here is that the implicit conversion to `void*` is disallowed:
```csharp
// C# errors today, despite void* being the "super-type" of all pointers
void* x = &Foo;

// C# allows this for `static void Foo()`
void* x = (delegate*<void>)&Foo;
```

I agree that allowing direct casting to any delegate signature is potentially error prone and likely undesirable. But requiring the cast to be present for what will be an implicit conversion to the super-type seems a bit odd

**Reply by PathogenDavid** [reply]
> ```cs
> // C# errors today, despite void* being the "super-type" of all pointers
> void* x = &Foo;
> ```

I went to dig this up only to remember you were part of the conversation 😅 For everyone else, here's some background on why this isn't allowed: https://github.com/dotnet/roslyn/issues/44489#issuecomment-632253853

TL;DR: Because it would be ambiguous if `Foo` was overloaded and would make adding a non-conflicting overload a breaking change.

Although that reasoning is somewhat moot after [C# 10's lambda improvements made `var x = Foo;` legal](https://github.com/dotnet/csharplang/blob/3fdda469fecc71dac96a93663e73f165ca16c3a3/proposals/csharp-10.0/lambda-improvements.md#var).

**Reply by tannergooding** [reply]
> I went to dig this up only to remember you were part of the conversation 😅

If only I could remember all the conversations I've been a part of  😆 

Now that I remember why it's this way, it makes sense. But it's also been annoying in practice... 

Maybe if `void* x = (void*)&Foo;` existed as a workaround it wouldn't be so annoying/unreadable (but there is the same general issue with picking the right overload still)

**Comment by jkotas** [comment, marked as answer]
> `(delegate*<uint, void>)(delegate*<int, void>)&Foo<int>;`

This example is a non-portable code. It assumes that `int` and `uint` arguments are passed exactly the same by the target architecture ABI. It happens to be the case for number of ABIs, but there are ABIs where it is not the case.

If the example is changed to cast between mismatched return values `(delegate*<ushort>)(delegate*<short>)&Foo`, it would not work correctly even on the existing CoreCLR Intel and Arm ABIs. (I just happened to root cause a codegen bug related to it a few hours ago.)

> Ok, but looks ugly

Brittle non-portable code looking ugly is a feature, not a bug.

**Reply by Sergio0694** [reply]
I see, that's fair. I guess Tanner and I were mostly confused because the current behavior is at least not "expected", in that anywhere else in the language you're free to cast the result of an expression returning a pointer of type `TU*` to one of type `TV*` without having to explicitly insert a cast to `TU*` first. Given this also applies to that (it just happens that `TU*` is of type `<SOME_DELEGATE>*`) this is the only case I can think of where the language isn't consistent with respect to this, so it seemed odd. But yeah I can see the argument for this working the way it does being a good thing now. Thanks! 🙂

**Reply by tannergooding** [reply]
> but there are ABIs where it is not the case.

Could you share some examples? This would likely be good to document examples of for users writing interop code and potentially as something handled by the DLL Import Generator. Many users like to use signed or other types rather than matching the actual native signature and if we currently support (or will support or Mono supports) such platforms where that's incorrect, it'd be good to ensure they are covered.

**Reply by jkotas** [reply]
> Could you share some examples? 

[Risc-V](https://riscv.org/wp-content/uploads/2015/01/riscv-calling.pdf) and [LoongArch64](https://github.com/loongson/LoongArch-Documentation/blob/main/docs/LoongArch-ELF-ABI-EN.adoc#c-data-types-and-alignment) are the two I know about.

**Reply by tannergooding** [reply]
Ah, I see. They both enforce zero vs sign-extension of the data 8/16-bit data and so if the actual method takes `short` and you call it using `ushort`, the value may be misinterpreted.

**Reply by jkotas** [reply]
> They both enforce zero vs sign-extension of the data 8/16-bit data

And 32-bit data too when the system is 64-bit.

**Comment by ghost** [comment]
```
    public static class Program
    {
        public static unsafe void Main()
        {
            MyAction action1 = (delegate*<void>)&Main; // IDE0004: Cast is redundant.
            MyAction action2 = &Main; // CS8812: Cannot convert &method group 'Main' to non-function pointer type 'MyAction'.
        }
    }

    public unsafe struct MyAction
    {
        public MyAction(delegate*<void> ptr)
        {
            this.ptr = ptr;
        }

        private readonly delegate*<void> ptr;

        public static implicit operator MyAction(delegate*<void> ptr)
        {
            return new MyAction(ptr);
        }
    }
```

**Reply by ghost** [reply]
```
MyAction action1 = (delegate*<void>)&Main; // IDE0004: Cast is redundant.
MyAction action2 = &Main; // CS8812: Cannot convert &method group 'Main' to non-function pointer type 'MyAction'.
```
Why is the cast of ``(delegate*<void>)&Main`` redundant, if ```&Main``` couldn’t implicitly convertible to ```delegate*<void>```?

**Reply by PathogenDavid** [reply]
This is a bug in IDE0004 that's been fixed but the fix hasn't rolled out yet. If it's driving you crazy, I did confirm it's fixed in the latest Visual Studio 2022 preview.

See discussion here: https://github.com/dotnet/roslyn/issues/58718
